## Lets start by reading the data, and tokenizing it by the help of `nltk`

In [1]:
import nltk
nltk.download('punkt') # to enable tokenizer of nltk
from nltk.tokenize import sent_tokenize, word_tokenize 
import numpy as np

[nltk_data] Downloading package punkt to /home/salil/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [2]:
dataset_name = "shakespeare.txt"
raw_data = open(dataset_name, "r").read().lower() # lets make the whole text to lower case as it will not differenciate b/w The and the

In [3]:
print("Total number of letters:", len(raw_data))
print("Total number of words:",len(raw_data.split(' ')))
print("Total number of letter in these each words:",end = '')
s = 0
for i in raw_data.split(' '):
    s+=len(i)
print('',s)

Total number of letters: 4194411
Total number of words: 691273
Total number of letter in these each words: 3503139


In [9]:
tokens = word_tokenize(raw_data)
print("Initially tokens variable is:",type(tokens),"lets convert it into numpy array")
tokens = np.asarray(tokens)
uniqTokens = np.unique(tokens)
print("#Tokens:",len(tokens))
print("#Unique tokens:",len(uniqTokens))

Initially tokens variable is: <class 'list'> lets convert it into numpy array
#Tokens: 932696
#Unique tokens: 53433


# Reading the corporus task is complete. 
# Now we have to encode the taken tokens in lexographic order, we have below 2 ways: 
- One-hot encoding
- Word embedding
## One-hot encdoding: 
Each word will be represented by a vector of dimensions: `53433`(*no of unique tokens*), and encoding 932696 tokens into such vector will be memory heavy (**932696\*53433 integers: which roughly equal to 185GB, so we wont do that. As I cant offord 185GB of RAM and I am not in a mood to process and save each of those tokens one-by-one**)
## Word embedding
Each token will be encoded into maybe 100D vector so max-to-max *88MB* that is way lesser than 